### PR TITLE
[Fix]Fix the logic of pseudo bboxes predicted by teacher model in SemiBase…

### PR DESCRIPTION
**Pull request by Czm369**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Fix the logic of pseudo bboxes predicted by teacher model in SemiBaseDetector.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/models/detectors/semi_base.py [hunk 1]
```diff
@@ -76,12 +76,13 @@ def loss(self, multi_batch_inputs: Dict[str, Tensor],
         losses = dict()
         losses.update(**self.loss_by_gt_instances(
             multi_batch_inputs['sup'], multi_batch_data_samples['sup']))
-        origin_pseudo_instances, batch_info = self.get_pseudo_instances(
-            multi_batch_inputs['unsup_teacher'],
-            multi_batch_data_samples['unsup_teacher'])
+        with torch.no_grad():
```
Comment (ZwwWayne) [inline mmdet/models/detectors/semi_base.py:79]: why change this, we can keep using decorator

File: mmdet/models/detectors/semi_base.py [hunk 2]
```diff
@@ -140,16 +141,36 @@ def loss_by_pseudo_instances(self,
         return rename_loss_dict('unsup_',
                                 reweight_loss_dict(losses, unsup_weight))
 
-    @torch.no_grad()
     def get_pseudo_instances(
```
Comment (ZwwWayne) [inline mmdet/models/detectors/semi_base.py:144]: this make the base class can only train with two-stage detector

File: mmdet/models/detectors/semi_base.py [hunk 3]
```diff
@@ -76,12 +76,13 @@ def loss(self, multi_batch_inputs: Dict[str, Tensor],
         losses = dict()
         losses.update(**self.loss_by_gt_instances(
             multi_batch_inputs['sup'], multi_batch_data_samples['sup']))
-        origin_pseudo_instances, batch_info = self.get_pseudo_instances(
-            multi_batch_inputs['unsup_teacher'],
-            multi_batch_data_samples['unsup_teacher'])
+        with torch.no_grad():
+            origin_pseudo_data_samples, batch_info = self.get_pseudo_instances(
```
Comment (hhaAndroid) [inline mmdet/models/detectors/semi_base.py:80]: The function name seems unreasonable, `get_pseudo_instances` actually returns datasample. The same  API deals with datasample of `project_pseudo_instances` and `loss_by_pseudo_instances`

File: mmdet/models/detectors/semi_base.py [hunk 4]
```diff
@@ -140,16 +141,36 @@ def loss_by_pseudo_instances(self,
...
+                x, batch_data_samples, rescale=False)
+
+        batch_info = {}
         for data_samples, results in zip(batch_data_samples, results_list):
-            data_samples.gt_instances = results.pred_instances
-        batch_info = None
+            data_samples.gt_instances = results
+            data_samples.gt_instances.bboxes = bbox_project(
+                data_samples.gt_instances.bboxes,
+                torch.from_numpy(data_samples.homography_matrix).inverse().to(
```
Comment (hhaAndroid) [inline mmdet/models/detectors/semi_base.py:171]: Why is homography_matrix not a torch.tensor object

**Review by hhaAndroid — APPROVED** [review]

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**